### PR TITLE
Offloaded KV Cache

### LINKED ISSUE #30704: KV cache with CPU offloading
### Feature request

I would like to contribute a KV cache implementation that only keeps a couple of layers on the GPU: the current layer in the forward pass as well as prefetching the next layer. The KV cache for the rest of the layers are on the CPU which usually has much larger capacity. Once the forward pass is done with a layer, the corresponding KV cache is evicted back to the CPU. It will be fetched again during the generation of the next token. This is all implemented in [this gist](https://gist.github.com/n17s/c0c6251ecb569a0c8efb0b0ec4b59e15) which can be used as a drop-in replacement for the `transformers.cache_utils.DynamicCache` class.

### Motivation

Performing inference with large language models on very long contexts can easily run out of GPU memory. The vast majority of the memory is consumed by the KV cache. The KV cache is indexed by layer and accessed sequentially by layer index during the forward pass. This gives us both a natural segmentation of the KV cache by layer index as well as a clear pattern for prefetching and eviction. When we do the forward pass on layer k, we evict layer k-1 and prefetch layer k+1. Thus, there are always at most 2 layers of KV cache in GPU memory.

The overlap of I/O and compute is not perfect. For example, for a 7b parameter model on a single H100 I am getting 12 tokens/sec with my KV cache vs. 16 tokens/sec with the standard KV cache for a context size of 8k tokens. However, the standard KV cache goes out of memory at a context size of 128k tokens while mine still works.

### Your contribution

I have provided an initial implementation in [this gist](https://gist.github.com/n17s/c0c6251ecb569a0c8efb0b0ec4b59e15). Not sure what is the right way to integrate it. Could be a new class, or an option for DynamicCache, or an automatic fallback when we are about to go out of memory, not sure. I'm also not very familiar with the PR process, since this is my first issue, so maybe it would be better if someone from the HF team can shepherd this through.

### PR DESCRIPTION
**Pull request by n17s**
# What does this PR do?

Fixes #30704 

This PR introduces OffloadedCache. This is a KV cache implementation that reduces GPU memory usage in exchange for more CPU memory usage and a small increase in generation time. During the forward passes in `generate`, it only keeps two layers of KV cache on the device: the current layer and the next layer. All other layers are on the CPU and are prefetched/evicted as necessary.

It can be used by passing `cache_implementation="offloaded"` in the `GenerationConfig` like this:
```python
        gen_config = GenerationConfig(
            cache_implementation="offloaded", 
            # other generation options such as 
            num_beams=4,
            num_beam_groups=2,
            num_return_sequences=4,
            diversity_penalty=1.0,
            max_new_tokens=50,
            early_stopping=True,
        )

        outputs = model.generate(
            inputs["input_ids"],
            generation_config=gen_config
        )
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. Issue #30704
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR. @ArthurZucker @gante

### CODE REVIEW

File: src/transformers/cache_utils.py [hunk 1]
```diff
@@ -447,6 +447,123 @@ def batch_select_indices(self, indices: torch.Tensor):
...
+        return cache
+
+    def to_legacy_cache(self) -> Tuple[Tuple[torch.Tensor], Tuple[torch.Tensor]]:
+        """Converts the `OffloadedCache` instance into the its equivalent in the legacy cache format."""
+        legacy_cache_lst = []
+        for layer_idx in range(len(self)):
+            original_device = self.original_device[layer_idx]
+            self.key_cache[layer_idx] = self.key_cache[layer_idx].to(original_device)
+            legacy_cache_lst.append((self.key_cache[layer_idx], self.value_cache[layer_idx]))
+        return tuple(legacy_cache_lst)
```
Comment (gante) [inline src/transformers/cache_utils.py:564]: I would overload these two functions with a `NotImplementedError` exception -- they are exclusively used for backwards compatibility, and we would like to get rid of them in the future 🤗

**Review by gante — COMMENTED** [review]
@n17s The core of the PR LGTM, thank you for proposing and implementing this cool feature 💛 

To finalize this PR:
1. Let's add an integration test (see [this file](https://github.com/huggingface/transformers/blob/main/tests/test_cache_utils.py) for examples). In particular, I would test that a) using this cache yields the same results as the original `DynamicCache`; b) if we're on a GPU, we should see a lower memory consumption than with `DynamicCache`
2. If you have benchmarks (and a benchmark script), please share it in this PR for future reference!

File: src/transformers/cache_utils.py (hunk 1 above)
Comment (n17s) [reply src/transformers/cache_utils.py]: I have removed these implementations but did it the way it is recommended in the Python docs and left a comment in the code with a pointer to the docs.

**Comment by n17s** [comment]
I have incorporated the comment about removing the two legacy implementations and added a couple of tests that ensure the cache works the same way as the DynamicCache and its peak memory usage is lower.

**Comment by n17s** [comment]
Is something holding this back? The failing tests are all marked as flaky.

File: tests/test_cache_utils.py [hunk 2]
```diff
@@ -455,3 +456,58 @@ def test_static_cache_extra_left_padding(self):
...
     def test_static_cache_beam_search(self):
         pass
+
+    @require_torch_gpu
+    def test_offloaded_cache_equivalent_to_dynamic_cache(self):
+        """Tests that OffloadedCache produces the same result as the default DynamicCache"""
+        model_name = "microsoft/Phi-3-mini-4k-instruct"
+        tokenizer = AutoTokenizer.from_pretrained(model_name)
+        model = AutoModelForCausalLM.from_pretrained(
+            model_name, device_map="auto", torch_dtype=torch.float16, attn_implementation="flash_attention_2"
```
Comment (gante) [inline tests/test_cache_utils.py:466]:
```suggestion
            model_name, device_map="auto", torch_dtype=torch.float16
```
FA2 places additional requirements on the test. Given that it is not relevant for the test itself, we should remove it 🤗

**Review by gante — APPROVED** [review]
LGTM, thank you for iterating 🤗

File: tests/test_cache_utils.py (hunk 2 above)
Comment (n17s) [reply tests/test_cache_utils.py]: I have removed flash attention from the tests.

File: tests/test_cache_utils.py [hunk 3]
```diff
@@ -455,3 +456,58 @@ def test_static_cache_extra_left_padding(self):
...
+        for original_output, offloaded_output in zip(original_outputs, offloaded_outputs):
+            assert torch.all(original_output == offloaded_output).item()
+
+    @require_torch_gpu
+    def test_offloaded_cache_uses_less_memory_than_dynamic_cache(self):
+        """Tests that OffloadedCache uses less memory than the default DynamicCache"""
+        model_name = "microsoft/Phi-3-mini-4k-instruct"
+        tokenizer = AutoTokenizer.from_pretrained(model_name)
+        model = AutoModelForCausalLM.from_pretrained(
+            model_name, device_map="auto", torch_dtype=torch.float16, attn_implementation="flash_attention_2"
```
[the remark this answers is not in the document]
Comment (n17s) [reply tests/test_cache_utils.py]: Fixed.

**Comment by n17s** [comment]
> just the limited bandwidth on our end

@gante Oh sorry, I was not aware. I removed flash attention but this unmasked a bug related to synchronization of my non-blocking moving of tensors (never tested without it before😁). I have added explicit synchronization right before a tensor is needed on the device or is safe to be evicted. This makes the tests pass while still keeping all the moving non-blocking. 

Some tests are still failing but they seem unrelated and/or flaky.

**Comment by n17s** [comment]
Since the tests have moved, I have rebased to the latest.

**Comment by n17s** [comment]
Gentle ping @ArthurZucker

**Comment by ArthurZucker** [comment]
I was off for a week! Back now, will review!

File: src/transformers/cache_utils.py [hunk 4]
```diff
@@ -447,6 +447,110 @@ def batch_select_indices(self, indices: torch.Tensor):
...
+    def __init__(self) -> None:
+        super().__init__()
+        self.original_device = []
+        self.prefetch_stream = torch.cuda.Stream()
+        self.beam_idx = None  # used to delay beam search operations
+
+    def prefetch_layer(self, layer_idx: int):
+        "Starts prefetching the next layer cache"
+        if layer_idx < len(self):
+            with torch.cuda.stream(self.prefetch_stream):
```
Comment (ArthurZucker) [inline src/transformers/cache_utils.py:467]: does that only work when using compile or in general ?

File: src/transformers/cache_utils.py [hunk 5]
```diff
@@ -447,6 +447,110 @@ def batch_select_indices(self, indices: torch.Tensor):
             self.value_cache[layer_idx] = self.value_cache[layer_idx][indices, ...]
 
 
+class OffloadedCache(DynamicCache):
+    """
```
Comment (ArthurZucker) [inline src/transformers/cache_utils.py:451]: at some point we should warn if generating on CPU that GPU is required + if we use cuda stream this requires a hardware that supports cuda streams

File: src/transformers/generation/utils.py [hunk 6]
```diff
@@ -1794,6 +1795,8 @@ def generate(
                     )
 
                 model_kwargs["past_key_values"] = cache_class(cache_config)
+            elif generation_config.cache_implementation == "offloaded":
+                model_kwargs["past_key_values"] = OffloadedCache()
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1799]: why don't we update the mapping that gives us the cahce class here?

**Review by ArthurZucker — COMMENTED** [review]
Looks great, small Q about using cuda streams, we should also add this to the doc / show and example as I guess this works with compile no?

File: src/transformers/cache_utils.py (hunk 4 above)
Comment (n17s) [reply src/transformers/cache_utils.py:467]: CUDA streams should work in general. torch.compile docs don't mention any known issues related to streams.

File: src/transformers/cache_utils.py (hunk 5 above)
Comment (n17s) [reply src/transformers/cache_utils.py:451]: Hmm, the point of this class is you want to generate on the GPU but you have more CPU RAM than GPU RAM. If you are generating on the CPU you don't need this class at all.

File: src/transformers/generation/utils.py (hunk 6 above)
Comment (n17s) [reply src/transformers/generation/utils.py:1799]: There is no such mapping? The only mappings I see in the code are
```python
 NEED_SETUP_CACHE_CLASSES_MAPPING = {"static": StaticCache, "sliding_window": SlidingWindowCache, "hybrid": HybridCache}
 QUANT_BACKEND_CLASSES_MAPPING = {"quanto": QuantoQuantizedCache, "HQQ": HQQQuantizedCache}
```
OffloadedCache does not need setup and is not a quantized cache.

**Comment by n17s** [comment]
> Looks great, small Q about using cuda streams, we should also add this to the doc / show and example as I guess this works with compile no?

Not sure. How to test it?

File: src/transformers/generation/utils.py (hunk 6 above)
Comment (ArthurZucker) [reply src/transformers/generation/utils.py:1799]: Yeah I was thinking about the `NEED_SETUP_CACHE_CLASSES_MAPPING`, but alrgith

**Review by ArthurZucker — COMMENTED** [review]
You can just compile the model's forward and do this: 
```
import os
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
os.environ["TOKENIZERS_PARALLELISM"] = "false" # silence warnings when compiling

device = "cuda"
ckpt = "model-to-test"

model = AutoModelForCausalLM.from_pretrained(ckpt, torch_dtype=torch.float16)
model.to(device)

tokenizer = AutoTokenizer.from_pretrained(ckpt)

prompt = "Why dogs are so cute?"
inputs = tokenizer(prompt, return_tensors="pt").to("cuda")

# Specify the max length (including both the prompt and the response)
# When calling `generate` with `cache_implementation="static" later, this is also used to create a `StaticCache` object
# with sequence length = `max_length`. The longer the more you will re-use it
model.generation_config.max_length = 128

# without `torch.compile`: each call takes ~ 5.0 seconds (on A100 80G + torch 2.3)
outputs = model.generate(**inputs, do_sample=False)
response = tokenizer.batch_decode(outputs)[0]
print(response)

# `torch.compile(model, ...)` is not recommended as you compile callbacks
# and full generate. We recommend compiling only the forward for now. 
# "reduce-overhead" will use cudagraphs. 
model.forward = torch.compile(model.forward, mode="reduce-overhead", fullgraph=True)
model.generation_config.cache_implementation = "static"

# with `torch.compile` (on A100 80G + torch 2.3)
# 1st call: ~ 90 seconds
outputs = model.generate(**inputs, do_sample=False)
response = tokenizer.batch_decode(outputs)[0]
# 2nd call: ~ 60 seconds
outputs = model.generate(**inputs, do_sample=False)
response = tokenizer.batch_decode(outputs)[0]
# 3nd call: ~ 1.5 seconds
outputs = model.generate(**inputs, do_sample=False)
response = tokenizer.batch_decode(outputs)[0]
print(response)
```
What is missing here for me is mostly raise error to protect usage (on GPU) mostly, resolve merge conflicts and should be good to go

**Comment by n17s** [comment]
> You can just compile the model's forward and do this:

So this does not work with offloaded cache. The error message is 
```
AttributeError: 'NoneType' object has no attribute 'synchronize'
```
Stemming from
```python
    self.prefetch_stream.synchronize()
```

Based on my understanding of the discussion on [this pytorch issue](https://github.com/pytorch/pytorch/issues/92804) what is happening is that CUDA streams are not preserved by the current implementation of torch.compile. Since this seems to be a limitation with how torch.compile currently handles things, and may eventually be fixed according to the discussion, I'd say there's not much that can be done on our side.

**Comment by n17s** [comment]
> What is missing here for me is mostly raise error to protect usage (on GPU) mostly, resolve merge conflicts and should be good to go

Merge conflicts have been resolved and I added a check in the constructor to protect usage. Fun fact, the class works even if inference happens on the CPU, as long as there's a GPU on the machine.

File: src/transformers/cache_utils.py (hunk 5 above)
Comment (n17s) [reply src/transformers/cache_utils.py:451]: Added a check to protect this.

**Review by ArthurZucker — COMMENTED** [review]
Very nice! One thing missing is a bit of doc / doc with an exampe of how to easily activate this : ex with high beam search ! either in https://huggingface.co/docs/transformers/v4.43.2/en/generation_strategies or in https://huggingface.co/docs/transformers/en/perf_infer_gpu_one !

**Comment by ArthurZucker** [comment]
(It's agreen on my side to merge otherwise!)

**Comment by n17s** [comment]
> One thing missing is a bit of doc / doc with an exampe of how to easily activate this
Added some docs

File: src/transformers/cache_utils.py [hunk 7]
```diff
@@ -447,6 +447,111 @@ def batch_select_indices(self, indices: torch.Tensor):
...
+    A drop-in replacement for DynamicCache that conserves GPU memory at the expense of more CPU memory.
+    Useful for generating from models with very long context.
+
+    When layer k is executing it moves the cache of layer k-1 to the CPU and prefetches the KV cache of layer k+1.
+    """
```
Comment (ArthurZucker) [inline src/transformers/cache_utils.py:456]: would make sense to explain how the streams work, basically one stream for prefetching, and one stream that computes

File: src/transformers/cache_utils.py [hunk 8]
```diff
@@ -447,6 +447,111 @@ def batch_select_indices(self, indices: torch.Tensor):
...
+                device = self.original_device[layer_idx]
+                self.key_cache[layer_idx] = self.key_cache[layer_idx].to(device, non_blocking=True)
+                self.value_cache[layer_idx] = self.value_cache[layer_idx].to(device, non_blocking=True)
+
+    def evict_previous_layer(self, layer_idx: int):
+        "Moves the previous layer cache to the CPU"
+        if len(self) > 2:
+            prev_layer_idx = (layer_idx - 1) % len(self)
+            self.key_cache[prev_layer_idx] = self.key_cache[prev_layer_idx].to("cpu", non_blocking=True)
+            self.value_cache[prev_layer_idx] = self.value_cache[prev_layer_idx].to("cpu", non_blocking=True)
```
Comment (ArthurZucker) [inline src/transformers/cache_utils.py:479]: can't the prefetch stream be used for that as well?

**Review by ArthurZucker — COMMENTED** [review]
Very nice!

File: src/transformers/cache_utils.py (hunk 8 above)
Comment (n17s) [reply src/transformers/cache_utils.py:479]: It should not! Eviction needs to be queued after all other operations on these tensors happening on the default stream. If we call it from prefetch stream, it is possible that the prefetch stream moves it out (and the memory allocator puts something else in that memory) before all read operations from the default stream are done. I will add a small comment in the code about this.

File: src/transformers/cache_utils.py [hunk 9]
```diff
@@ -447,6 +447,111 @@ def batch_select_indices(self, indices: torch.Tensor):
...
+class OffloadedCache(DynamicCache):
+    """
+    A drop-in replacement for DynamicCache that conserves GPU memory at the expense of more CPU memory.
+    Useful for generating from models with very long context.
+
+    When layer k is executing it moves the cache of layer k-1 to the CPU and prefetches the KV cache of layer k+1.
+    """
+
+    def __init__(self) -> None:
+        assert torch.cuda.is_available(), "OffloadedCache can only be used with a GPU"
```
[the remark this answers is not in the document]
Comment (n17s) [reply src/transformers/cache_utils.py]: Fixed

File: src/transformers/cache_utils.py (hunk 7 above)
Comment (n17s) [reply src/transformers/cache_utils.py:461]: Done

**Review by ArthurZucker — APPROVED** [review]
Nice thanks for iterating!

**Comment by ghadiaravi13** [comment]
this cache is very helpful (for low-memory long-context inference)
thanks for adding it!

**Pull request merged**